### PR TITLE
Migrate travis to github action

### PR DESCRIPTION
**Pull request by liya2017**
Migrate travis to github action
TODO:
1. Need test for PR merge by bors
2. Refactor and test windows download secret.asc 
3. Need move the bors file to the bors config file

### CODE REVIEW

File: .github/workflows/pull_request_verify.yaml [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+        -  mac-runner
+    steps:
+    - name: Set env
+      run: echo "CARGO_TARGET_DIR=$(echo $HOME/target)" >> $GITHUB_ENV
+    - uses: actions/checkout@v2
+      with:
+        submodules: false
+        fetch-depth: 2
+        ref: ${{ github.event.pull_request.head.sha }}
+        repository: ${{github.event.pull_request.head.repo.full_name}}
```
Comment (keroro520) [inline .github/workflows/pull_request_verify.yaml:31]: I think it's OK that uses the default inputs.

File: .github/workflows/pull_request_verify.yaml [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+        #cargo build 的target，试一下在job 层加 cargo build targe
+    - run: |
+        echo "runner os is ${{ runner.os }}"
+        if [ ${{ runner.os }} == "macOS" ]; then
+        export PATH="/Users/administrator/.cargo/bin:$PATH"
```
Comment (keroro520) [inline .github/workflows/pull_request_verify.yaml:36]: https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#adding-a-system-path

File: .github/workflows/pull_request_verify.yaml [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        fi
+        echo $PATH
```
Comment (keroro520) [inline .github/workflows/pull_request_verify.yaml:38]: remove staled echo message.

File: .github/workflows/pull_request_verify.yaml [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+    needs: Test
+    steps:
+    - name: Set env
+      run: echo "CARGO_TARGET_DIR=$(echo $HOME/target)" >> $GITHUB_ENV
+    - name: Linters
+      run: |
+         if [ ${{ runner.os }} == "macOS" ]; then
+         export PATH="/Users/administrator/.cargo/bin:$PATH"
+         fi
+         cargo fmt --version || travis_retry rustup component add rustfmt
```
Comment (keroro520) [inline .github/workflows/pull_request_verify.yaml:74]: `travis_retry` ?

File: .github/workflows/pull_request_verify.yaml [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        -  Linux-runner
+        -  mac-runner
+    needs: Test
+    steps:
+    - name: Set env
+      run: echo "CARGO_TARGET_DIR=$(echo $HOME/target)" >> $GITHUB_ENV
+    - name: Linters
+      run: |
+         if [ ${{ runner.os }} == "macOS" ]; then
+         export PATH="/Users/administrator/.cargo/bin:$PATH"
```
Comment (keroro520) [inline .github/workflows/pull_request_verify.yaml:72]: fix the indentation

File: .github/workflows/pull_request_verify.yaml [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+    needs: Linters 
+    continue-on-error: true
+    steps:
+    - name: Set env
+      run: echo "CARGO_TARGET_DIR=$(echo $HOME/target)" >> $GITHUB_ENV
+    - name: Latest_Linters
+      run: |
+        if [ ${{ runner.os }} == "macOS" ]; then
+        export PATH="/Users/administrator/.cargo/bin:$PATH"
+        rustup component add rustfmt-preview
```
Comment (keroro520) [inline .github/workflows/pull_request_verify.yaml:95]: I would like to suggest that move these pre-install toolchains into a single step or a single job, and other jobs dependend this pre-install job.

File: .github/workflows/pull_request_verify.yaml [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+        rustup component add clippy-preview
+        fi
+        if [ ${{ runner.os }} == "Linux" ]; then
+        rustup component add rustfmt --toolchain stable-x86_64-unknown-linux-gnu
+        rustup component add clippy --toolchain stable-x86_64-unknown-linux-gnu
+        fi
+        mv rust-toolchain rust-toolchain.bak
+        echo "stable" > rust-toolchain
```
Comment (keroro520) [inline .github/workflows/pull_request_verify.yaml:103]: why `echo "stable" > rust-toolchain` ?

**Review by keroro520 — CHANGES_REQUESTED** [review]

File: .github/workflows/pull_request_verify.yaml [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+  CKB_CLI_VERSION: v0.40.0  
+jobs:
+#job for checkout the ckb
+  checkout: 
+    name: ckb-checkout  
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os:
+        -  Linux-runner
```
[the remark this answers is not in the document]
Comment (liya2017) [reply .github/workflows/pull_request_verify.yaml:21]: sure, it's just a label name, we can change it

File: .github/workflows/pull_request_verify.yaml [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+        -  mac-runner
+    steps:
+    - name: Set env
+      run: echo "CARGO_TARGET_DIR=$(echo $HOME/target)" >> $GITHUB_ENV
```
[the remark this answers is not in the document]
Comment (liya2017) [reply .github/workflows/pull_request_verify.yaml:25]: no, if I set it by env, the $home is unrecognized.

File: .github/workflows/pull_request_verify.yaml (hunk 3 above)
Comment (liya2017) [reply .github/workflows/pull_request_verify.yaml:38]: done

File: .github/workflows/pull_request_verify.yaml (hunk 4 above)
Comment (liya2017) [reply .github/workflows/pull_request_verify.yaml:74]: sorry, I forget there, and I am thing about to retry by github actions

File: .github/workflows/pull_request_verify.yaml (hunk 7 above)
Comment (liya2017) [reply .github/workflows/pull_request_verify.yaml:103]: I am not sure, just migrate from travis

File: .github/workflows/rc_branch.yaml [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+        include:
+          - os: Linux-runner
+            script_run: RUST_BACKTRACE=1 RUST_LOG=${INTEGRATION_RUST_LOG} test/run.sh --release -- --bin $CARGO_TARGET_DIR/release/ckb ${CKB_TEST_ARGS}  
+          - os:  mac-runner
+            script_run: export PATH="/Users/administrator/.cargo/bin:$PATH" && RUST_BACKTRACE=1 RUST_LOG=${INTEGRATION_RUST_LOG} test/run.sh --release -- --bin $CARGO_TARGET_DIR/release/ckb ${CKB_TEST_ARGS}
+          - os: win-runner
+            script_run: devtools/windows/make CKB_TEST_SEC_COEFFICIENT=5 CKB_TEST_ARGS="-c 4 --no-report" integration 
+            CI: true
+            SENTRY_DSN: 'https://15373165fbf2439b99ba46684dfbcb12@sentry.nervos.org/7'
+            LOGBAK_USER: $(LOGBAK_USER)
```
Comment (keroro520) [inline .github/workflows/rc_branch.yaml:68]: That's just my confusion.

Could you tell me what `$(LOGBAK_*)` means?
Have you ever tested this job for "win-runner"?

**Review by keroro520 — CHANGES_REQUESTED** [review]

File: .github/workflows/package.yaml [hunk 11]
```diff
@@ -0,0 +1,57 @@
+name: Package
+
+# When someone comment "/deploy" on anyone PR,
+# this workflow will be triggered and run.
```
Comment (keroro520) [inline .github/workflows/package.yaml:4]: Incorrect comment

File: .github/workflows/package.yaml [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+on:
+  push:
+    branches:
+      - 'pkg/*'
+
+env:
+  CARGO_TERM_COLOR: always
+  RUST_BACKTRACE: full
+  CKB_CLI_VERSION: v0.40.0
+  GITHUB_TOKEN: "${{ secrets.GITHUB_TOKEN }}"
```
Comment (keroro520) [inline .github/workflows/package.yaml:14]: Where we use `env.GITHUB_TOKEN`?

**Review by keroro520 — CHANGES_REQUESTED** [review]
Suggestion: merge `cronjob_master` and `cronjob_develop`.

File: .github/workflows/rc_branch.yaml (hunk 10 above)
Comment (liya2017) [reply .github/workflows/rc_branch.yaml:63]: I am not sure what $(LOGBAK_*) means, it's the env var from azure. The integration test passed on win-runner, but I am trying to find away to make sure I can get value via this way.

Comment (liya2017) [reply .github/workflows/rc_branch.yaml:63]: it's seems uploda log.zip to azure

File: .github/workflows/pull_request_verify.yaml [hunk 13]
```diff
@@ -0,0 +1,191 @@
+name: Pull Requests
+
+# When someone comment "/deploy" on anyone PR,
+# this workflow will be triggered and run.
+on:
+  pull_request:
+    types: [ opened, synchronize, reopened, labeled ]
```
[the remark this answers is not in the document]
Comment (liya2017) [reply .github/workflows/pull_request_verify.yaml]: fixed

File: .github/workflows/pull_request_verify.yaml (hunk 1 above)
Comment (liya2017) [reply .github/workflows/pull_request_verify.yaml]: removed it, uses the default inputs.

File: .github/workflows/pull_request_verify.yaml (hunk 2 above)
Comment (liya2017) [reply .github/workflows/pull_request_verify.yaml]: added the step to set the path

File: .github/workflows/pull_request_verify.yaml (hunk 4 above)
Comment (liya2017) [reply .github/workflows/pull_request_verify.yaml]: removed it

File: .github/workflows/pull_request_verify.yaml (hunk 6 above)
Comment (liya2017) [reply .github/workflows/pull_request_verify.yaml]: add the pre step to install rustup component

File: .github/workflows/rc_branch.yaml (hunk 10 above)
Comment (liya2017) [reply .github/workflows/rc_branch.yaml]: Need keep those vars, will upload the log file when the test failed. but i have not test that part yet

File: .github/workflows/package.yaml (hunk 11 above)
Comment (liya2017) [reply .github/workflows/package.yaml]: fixed

File: .github/workflows/package.yaml (hunk 12 above)
Comment (liya2017) [reply .github/workflows/package.yaml]: removed it

File: .github/workflows/bors.toml [hunk 14]
```diff
@@ -0,0 +1,4 @@
+status = [ "ci" ]
+
+timeout_sec = 7200
+delete_merged_branches = true
```
Comment (keroro520) [inline .github/workflows/bors.toml:4]: Why is it in the workflows folder? Is it a Github Bot configuration file?

File: .github/workflows/cronjob.yaml [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+
+env:
+  CARGO_TERM_COLOR: always
+  RUST_BACKTRACE: full
+  CARGO_TARGET_DIR: ${{ github.workspace }}/../target
+  LOGBAK_USER: ${{secrets.LOGBAK_USER}}  #LOCBAK_* for upload logs to server when test failed, for windows
+  LOGBAK_PASSWORD: ${{secrets.LOGBAK_PASSWORD}}
+  LOGBAK_SERVER: ${{secrets.LOGBAK_SERVER}}
+jobs:
+#job for checkout the ckb
```
Comment (keroro520) [inline .github/workflows/cronjob.yaml:15]: remove this comment

File: .github/workflows/cronjob.yaml [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+        include:
+          - os: Linux-runner
+            script_run: make build
+          - os:  mac-runner
+            script_run: make build
+          - os: win-runner
+            script_run: devtools/windows/make prod
+    steps:
+    - name: Set Cargo path
+      run: echo "$HOME/.cargo/bin" >> $GITHUB_PATH
```
Comment (keroro520) [inline .github/workflows/cronjob.yaml:31]:
```shell
grep '.cargo/bin" >> $GITHUB_PATH' .github -r | wc -l
33
```

It is too noisy.
The Github-hosted runners already pre-install Cargo and add it into the system path;
Our self-hosted runners should do this as well.

You can do it by adding cargo into `actions-runner/.path`

File: .github/workflows/cronjob.yaml [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
+    - uses: actions/checkout@v2
+      with:
+        submodules: false
+        fetch-depth: 2
```
Comment (keroro520) [inline .github/workflows/cronjob.yaml:35]: can we use default parameters?

File: .github/workflows/cronjob.yaml [hunk 18]
```diff
@@ -0,0 +1,227 @@
...
+      matrix:
+        os:
+        -  Linux-runner
+        -  mac-runner
+    needs: UnitTest
+    steps:
+    - name: Set Cargo path
+      run: echo "$HOME/.cargo/bin" >> $GITHUB_PATH
+    - name: Benchmarks_Test
+      run: make bench-test
```
Comment (keroro520) [inline .github/workflows/cronjob.yaml:75]: break line

File: .github/workflows/cronjob.yaml [hunk 19]
```diff
@@ -0,0 +1,227 @@
...
+        -  Linux-runner
+        -  mac-runner
+    needs: Integration_Test
+    if: ${{github.ref == 'refs/heads/master'}}
+    steps:
+    - name: Set Cargo path
+      run: echo "$HOME/.cargo/bin" >> $GITHUB_PATH
+    - name: Linters
+      run: |
+         cargo fmt --version || travis_retry rustup component add rustfmt
```
Comment (keroro520) [inline .github/workflows/cronjob.yaml:114]: `travis_retry`?

**Review by keroro520 — CHANGES_REQUESTED** [review]

File: .github/workflows/pull_request_verify.yaml [hunk 20]
```diff
@@ -0,0 +1,172 @@
...
+          - os: win-runner
+            script_run: devtools/windows/make prod
+    steps:
+    - name: Set Cargo path
+      run: echo "$HOME/.cargo/bin" >> $GITHUB_PATH
+    - uses: actions/checkout@v2
+      with:
+        submodules: false
+        fetch-depth: 2
+    - run: ${{ matrix.script_run }}
```
Comment (keroro520) [inline .github/workflows/pull_request_verify.yaml:36]: This job named "checkout". Then why it executes `run: ${{ matrix.script_run }}`?

Comment (keroro520) [reply .github/workflows/pull_request_verify.yaml:36]: And why "checkout" needs `strategy.matrix`

File: .github/workflows/pull_request_verify.yaml [hunk 21]
```diff
@@ -0,0 +1,172 @@
...
+        include:
+          - os: Linux-runner
+            script_run: make test
+          - os:  mac-runner
+            script_run: make test
+          - os: win-runner
+            script_run: devtools/windows/make test
+            CI: true
+            ImageOS: 'windows-2019'
+            BUILD_BUILDID: ${{ github.sha }}
```
Comment (keroro520) [inline .github/workflows/pull_request_verify.yaml:52]: There are some strategy matrices that specify `BUILD_BUILDID` and `ImageOS` , while others do not. Why is this?

File: .github/workflows/pull_request_verify.yaml [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+          - os: Linux-runner
+            script_run: make build
+          - os:  mac-runner
+            script_run: make build
+          - os: win-runner
+            script_run: devtools/windows/make prod
+    steps:
+    - name: Set Cargo path
+      run: echo "$HOME/.cargo/bin" >> $GITHUB_PATH
+    - uses: actions/checkout@v2
```
Comment (keroro520) [inline .github/workflows/pull_request_verify.yaml:32]: Use the default parameters

File: .github/workflows/pull_request_verify.yaml [hunk 23]
```diff
@@ -0,0 +1,172 @@
...
+            BUILD_BUILDID: ${{ github.sha }}
+    needs: checkout
+    steps:
+    - name: Set Cargo path
+      run: echo "$HOME/.cargo/bin" >> $GITHUB_PATH
+    - name: UnitTest
+      run: ${{ matrix.script_run }}
+    env:
+      CI: ${{ matrix.CI }}
+      SENTRY_DSN: ${{ matrix.SENTRY_DSN }}
```
Comment (keroro520) [inline .github/workflows/pull_request_verify.yaml:61]: What value is returned by `${{ matrix.SENTRY_DSN }}`? I did not see it declared.

File: .github/workflows/pull_request_verify.yaml [hunk 24]
```diff
@@ -0,0 +1,172 @@
...
+      ImageOS:  ${{ matrix.ImageOS }}
+
+  Linters:
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os:
+        -  Linux-runner
+        -  mac-runner
```
Comment (keroro520) [inline .github/workflows/pull_request_verify.yaml:70]: Same logic, but different style.

```
        os: [Linux-runner, mac-runner,win-runner]

```

```
      matrix:
        os:
        -  Linux-runner
        -  mac-runner
```

**Review by keroro520 — CHANGES_REQUESTED** [review]
The files pull_request_verify.yaml and merge_pr_by_bors.yaml are almost identical. Please merge them.

File: .github/workflows/cronjob.yaml (hunk 17 above)
Comment (keroro520) [reply .github/workflows/cronjob.yaml:35]:
```shell
grep -r submodules . | wc -l
5
```

File: .github/workflows/bors.toml (hunk 14 above)
Comment (liya2017) [reply .github/workflows/bors.toml:4]: yes, it's will go to Bot configuration file before the PR merge.

File: .github/workflows/cronjob.yaml (hunk 16 above)
Comment (liya2017) [reply .github/workflows/cronjob.yaml:31]: The ubuntu and windows don't need this , but mac need this.

File: .github/workflows/pull_request_verify.yaml (hunk 21 above)
Comment (liya2017) [reply .github/workflows/pull_request_verify.yaml:52]: those just for windows.

File: .github/workflows/pull_request_verify.yaml (hunk 20 above)
Comment (liya2017) [reply .github/workflows/pull_request_verify.yaml:36]: The checkout also include build steps, build windows and ubuntu/mas has different command.

File: .github/workflows/cronjob.yaml (hunk 17 above)
Comment (liya2017) [reply .github/workflows/cronjob.yaml:35]: I think it's better to keep the same config with travis.

File: .github/workflows/pull_request_verify.yaml (hunk 23 above)
Comment (liya2017) [reply .github/workflows/pull_request_verify.yaml:61]: fixed

File: .github/workflows/pull_request_verify.yaml (hunk 24 above)
Comment (liya2017) [reply .github/workflows/pull_request_verify.yaml:70]: fixed

File: .github/workflows/pull_request_verify.yaml (hunk 20 above)
Comment (liya2017) [reply .github/workflows/pull_request_verify.yaml:36]: removed build steps.

**Comment by liya2017** [comment]
bors r=keroro520

**Comment by liya2017** [comment]
bors r=keroro520

**Pull request closed without merging**